File: 08-tries/00208-Implement_Trie_Prefix_Tree/00208-implement_trie_prefix_tree.py
# 208: Implement Trie Prefix Tree
# https://leetcode.com/problems/implement-trie-prefix-tree


class TrieNode:
    def __init__(self) -> None:
        self.children: TrieNode[26] = []
        self.isWord = False

        for i in range(26):
            self.children.append(None)
            self.isWord = False

class Trie:
    def __init__(self):
        self.root: TrieNode = TrieNode()


    def insert(self, word: str) -> None:
        node: TrieNode = self.root
        current = 0;
        
        for i in range(len(word)):
            current = ord(word[i]) - ord('a')
            if node.children[current] == None:
                node.children[current] = TrieNode()
            node = node.children[current];

        node.isWord = True

    def search(self, word: str) -> bool:
        node: TrieNode = self.root
        current = 0;
        
        for i in range(len(word)):
            current = ord(word[i]) - ord('a')
            if node.children[current] == None:
                return False
            node = node.children[current];

        return node.isWord

    def startsWith(self, prefix: str) -> bool:
        node: TrieNode = self.root
        current = 0;
        
        for i in range(len(prefix)):
            current = ord(prefix[i]) - ord('a')
            if node.children[current] == None:
                return False
            node = node.children[current];

        return True
    

if __name__ == "__main__":
    trie: Trie = Trie()

    # OPERATIONS
    trie.insert("apple")
    print(trie.search("apple"))
    print(trie.search("app"))
    print(trie.startsWith("app"))
    trie.insert("app")
    print(trie.search("app"))

File: 08-tries/00208-Implement_Trie_Prefix_Tree/00208-implement_trie_prefix_tree.go
// 208: Implement Trie Prefix Tree
// https://leetcode.com/problems/implement-trie-prefix-tree

package main

import (
	"fmt"
)


type Trie struct {
    children [26]*Trie
	isWord bool
}


func Constructor() Trie {
	return Trie{}
}


func (this *Trie) Insert(word string)  {
    node := this
	current := 0
	
	for i:=0; i<len(word); i++ {
		current = int(word[i] - 'a')
		if node.children[current] == nil {
			node.children[current] = &Trie{}
		}
		node = node.children[current]
	}

	node.isWord = true
}


func (this *Trie) Search(word string) bool {
    node := this
	current := 0

	for i:=0; i<len(word); i++ {
		current = int(word[i] - 'a')
		if node.children[current] == nil {
			return false
		}
		node = node.children[current]
	}

	return node.isWord
}


func (this *Trie) StartsWith(prefix string) bool {
	node := this
	current := 0

	for i:=0; i<len(prefix); i++ {
		current = int(prefix[i] - 'a')
		if node.children[current] == nil {
			return false
		}
		node = node.children[current]
	}

	return true
}


func main() {
	trie := Constructor()

	// OPERATIONS
	trie.Insert("apple")
	fmt.Println(trie.Search("apple"))
	fmt.Println(trie.Search("app"))
	fmt.Println(trie.StartsWith("app"))
	trie.Insert("app")
	fmt.Println(trie.Search("app"))
}